editor/src/core/workers/parser-printer/uid-fix.ts

```diff
@@ -1,284 +1,569 @@
+  } else {
+    // In this case the uid needs updating.
+    return set(uidOptic, uidToUse, baseValue)
+  }
+}
+
+// Should we find the need for some smarter handling of arrays of things in the future.
+// Like for example handling something being inserted in the middle and shifting the uids appropriately.
+// Then improving this function will make all of the cases where it is used better.
+function fixArrayElements<T>(
+  fixElement: (oldElement: T, newElement: T) => T,
+  oldExpression: Array<T>,
+  newExpression: Array<T>,
+): Array<T> {
+  return newExpression.map((newElement, newElementIndex) => {
+    if (newElementIndex < oldExpression.length) {
+      const oldElement = oldExpression[newElementIndex]
+      return fixElement(oldElement, newElement)
+    } else {
+      return newElement
     }
   })
+}
 
-  return numberOfElementsMatch
+export function fixTopLevelElementsUIDs(
+  oldElements: Array<TopLevelElement>,
+  newElements: Array<TopLevelElement>,
+  fixUIDsState: FixUIDsState,
+): Array<TopLevelElement> {
+  return fixArrayElements(
+    (oldElement, newElement) => {
+      return fixTopLevelElementUIDs(oldElement, newElement, fixUIDsState)
+    },
+    oldElements,
+    newElements,
+  )
 }
 
-function walkElementsWithin(
-  pathSoFar: StaticElementPathPart,
-  oldElements: ElementsWithin,
-  newElements: ElementsWithin,
-  onElement: OnElement,
-): boolean {
-  const oldElementKeys = Object.keys(oldElements)
-  const newElementKeys = Object.keys(newElements)
-
-  const lengthsMatchAtThisLevel = newElementKeys.length === oldElementKeys.length
-  let numberOfElementsMatch = lengthsMatchAtThisLevel
-
-  newElementKeys.forEach((elementKey, index) => {
-    const newElement = newElements[elementKey]
-    const oldElementKey: string | null = oldElementKeys[index] ?? null
-    const oldElement: JSXElementChild | null =
-      oldElementKey == null ? null : oldElements[oldElementKey] ?? null
-
-    const lengthsOfChildrenMatch = compareAndWalkElements(
-      oldElement,
-      newElement,
-      pathSoFar,
-      onElement,
-    )
-    numberOfElementsMatch = numberOfElementsMatch && lengthsOfChildrenMatch
-  })
+export function fixTopLevelElementUIDs(
+  oldElement: TopLevelElement,
+  newElement: TopLevelElement,
+  fixUIDsState: FixUIDsState,
+): TopLevelElement {
+  switch (newElement.type) {
+    case 'UTOPIA_JSX_COMPONENT': {
+      if (oldElement.type === newElement.type) {
+        return fixUtopiaJSXComponentUIDs(oldElement, newElement, fixUIDsState)
+      }
+      break
+    }
+    case 'ARBITRARY_JS_BLOCK': {
+      if (oldElement.type === newElement.type) {
+        return fixArbitraryJSBlockUIDs(oldElement, newElement, fixUIDsState)
+      }
+      break
+    }
+    case 'IMPORT_STATEMENT': {
+      return newElement
+    }
+    case 'UNPARSED_CODE': {
+      return newElement
+    }
+    default:
+      const _exhaustiveCheck: never = newElement
+      throw new Error(`Unhandled case ${JSON.stringify(newElement)}.`)
```

Comment:
can this be replaced with `assertNever`?

Reply:
Fixed in 6a3d04e93.